src/main/java/filter/expression/Qualifier.java

```diff
@@ -136,28 +134,88 @@ public static boolean process(IModel model, FilterExpression expr, TurboIssue is
+     * i.e. 0 for "current" milestone, -1 for "current-1" milestone, etc.
+     */
+    private static Map<Integer, TurboMilestone> getMilestoneAliasIndex(List<TurboMilestone> milestones) {
+        List<TurboMilestone> sortedMilestones = TurboMilestone.sortByDueDate(milestones);
+        Optional<Integer> currentMilestoneIndex = getCurrentMilestoneIndex(sortedMilestones);
 
-        for (TurboMilestone checker : allMilestones) {
-            boolean overdue = checker.getDueDate().isPresent() &&
-                    checker.getDueDate().get().isBefore(LocalDate.now());
-            boolean relevant = !(overdue && checker.getOpenIssues() == 0);
+        assert currentMilestoneIndex.isPresent() || !currentMilestoneIndex.isPresent() && sortedMilestones.isEmpty();
 
-            if (checker.isOpen() && relevant) {
-                return Optional.of(currentIndex);
-            }
+        return IntStream
+                .range(0, sortedMilestones.size())
+                .boxed()
+                .collect(Collectors.toMap(
+                        i -> i - currentMilestoneIndex.get(),
+                        i -> sortedMilestones.get(i)));
+    }
 
-            currentIndex++;
+    /**
+     * "Current" milestone is an ongoing milestone with the earliest due date. However, if there
+     * is only one open milestone, it will be considered as the "current" milestone, even if it
+     * does not have due date.
+     *
+     * This method expects the milestone list to be sorted (by due date)
+     */
+    private static Optional<Integer> getCurrentMilestoneIndex(List<TurboMilestone> milestones) {
+        if (milestones.isEmpty()) {
+            return Optional.empty();
         }
 
-        // if no open milestone, set current as one after last milestone
-        // - this means that no such milestone, which will return no issue
-        return Optional.of(allMilestones.size());
+        int openMilestonesCount = TurboMilestone.getOpenMilestones(milestones).size();
+        if (openMilestonesCount == 1) {
+            return IntStream
+                    .range(0, milestones.size())
+                    .boxed()
+                    .filter(i -> milestones.get(i).isOpen())
+                    .findFirst();
+        } else {
+            // Look for the first milestone in the (sorted) list that is ongoing.
+            Optional<Integer> firstOngoingMilestonePosition = IntStream
+                    .range(0, milestones.size())
+                    .boxed()
+                    .filter(i -> milestones.get(i).isOngoing())
+                    .findFirst();
+
+            if (firstOngoingMilestonePosition.isPresent()) {
+                return firstOngoingMilestonePosition;
+            } else {
+                // if no ongoing milestone, set current as one after last milestone
+                // - this means that no such milestone, which will return no issue
+                return Optional.of(milestones.size());
+            }
+        }
```

Comment:
When the `if` branch exits, there is no need for `else`.
```java
if (openMilestonesCount == 1) {
    return getFirstOpenMilestonePosition(milestones);
}

Optional<Integer> firstOngoingMilestonePosition = getFirstOngoingMilestonePosition(milestones);

return firstOngoingMilestonePosition.isPresent()
           ? firstOngoingMilestonePosition 
           : Optional.of(milestones.size());
```

The logic is still not the most intuitive though.